### ISSUE TITLE
Review notes from Freya

### ISSUE THREAD

**Issue by cmcaine** [IntegratedTransportPlanning/gcvt#159]
re: @polycement

Quotes are from the initial review I asked Freya to make of the code. Other words are my responses.

> ## CODE:

> - inconsistent function definition style (const/function)
> - mix of " and '
> - mix of camelcase and underscores for var/func names

These are all slightly bad, but very minor and changing them now probably worse for making git diff / git blame harder to read

> - consider splitting this code into a few different files, shoving everything into a single 1400+ loc file where things like map are global does not smell great

Agree on the globals, maybe smaller files would be good too, generally Olie and I have preferred big files because they're easier to search in compared to project-wide search (e.g. not a fan of the one file per widget thing that lots of JS projects do, but maybe that's mostly just because of how we use vim).

I think it could be good to move the view and mapbox stuff out into separate files and clean them up a bit, tho.

Would make git blame unreadable, and there's a risk that it's just organisation for the sake of organisation.

Pls prioritise other work unless they become an issue for you.

> - stateFromSearch: qsObj can be const, first loop does nothing...?

Fixed.

let -> const

First loop is still there so we can add new floats easily, but JSON.parse also does that, so maybe unnecessary? I'll remove them for now

> - hardcoded mapboxgl access token smells bad but maybe unavoidable in frontend

I don't think this is a problem.

> - mapboxInit is 150+ lines long,

Yeah, it's a bit gnarly. Answered above re: splitting things into files.

> - line 395: prefer () over _ for argless arrow funcs

Fixed.

> - line 510: what's this semicolon doing here?

Leading semicolon before [ or ( is so that JS doesn't try to index into or call the expression on the preceding line. Not needed in this case, but if the line was copy pasted below it would need to be there.

> - i remember reading somewhere that forEach is better than for of, but i forget why so w/e

`for in` has some footguns, but `for of` is fine.

> - line 691: why is this bracketed? scope? why?

I think this is just organisational.

Fixed.

> - scenarios_with has what im like 99% sure is invalid if/else syntax and idk how this even runs (not called?)

The braces are optional if there's only one statement in the body of the if or else.

Fixed.

> - line 751: function def with leading uppercase...?

`Legend` is a mithril component, and those normally get uppercase names

> - menuView is like 180ish lines long...

Yeah, gnarly, but not sure what you could do to make it significantly nicer?

> ## UI/UX

> - copy link button text should probably have a shorter timer on switching back, i thought it'd broken

pls fix

> - region border lines on graph could do with some anti-aliasing

pls open an issue or fix. v low priority, tho

> - scenario & year seem to default to first in list but don't initially set their displayed values
> - toggling compare off and on again half loads what was set - feels like it should remember all or none of it
> - click to show flows tip vanishes while a region is selected (that might actually be my fault lol)
> - could probably use on-screen instructions for map controls (click, ctrl+click, etc)

pls investigate and fix these

> - comparing to self doesn't seem to work, should it even be allowed?

pls disable this

> - graph seems slower to update than i remember but that could just be me

Low priority / could be fake news

> - if content changes while tooltip is up over map, tooltip doesn't update til mouse is moved

yeah this has annoyed me for a while, but I haven't got around to fixing it. The issue is that we're not setting the tooltips in the view, really.

I even mention it in the comment at the top of the file :)

```
 * Notably, the popups that appear on hover and on click are mostly managed by
 * mapboxgl, which causes some minor problems.
```

Anyway, a fix would be to store the info needed to generate the popups in state and then regenerate them in the mapbox updater.

Would you be happy to investigate that and fix?

> - some boundaries don't sit flush with each other and seem generally misaligned, could just be bad data

It's bad data, not in scope for us.

> - centroids in flow view seem off but again could be bad data

Could be our fault or not.

> - selected region in flow view doesn't draw unless flows are toggled off...?

Don't know what you're talking about here.  See if you can fix it?

> - generally very strange behaviour when all flows are 0 for a region (eg. boryspil in kyiv set)

Dunno what to do here.

> - might be an idea to center chevrons relative to flow length, would lessen stacking when flows overlap

Could you try this and see? You need to look up the rules for "symbol-spacing" in the mapboxgljs style docs

> - clicking anywhere on selected regions in flow list deselects, not just the x on the left - is that ok?

I think this is fine? If you think it's bad UX pls fix

> - "click to show flows" still appears during a comparison, but clicking disables the comparison

Can you come up with a better UX for this, maybe just change the tooltip?

> - chevron colour on flows could maybe be better? its probably ok tho

Let's just leave it.

**Comment by cmcaine** [comment]
Thanks for looking at these. Could you fix some of them? I've ordered them in rough descending order of importance.

- [ ] copy link button text should probably have a shorter timer on switching back, i thought it'd broken

    pls fix

- [ ] centroids in flow view seem off but again could be bad data

    Could be our fault or not. Show me later in screenshot or at your PC?

- [ ] scenario & year seem to default to first in list but don't initially set their displayed values
- [ ] click to show flows tip vanishes while a region is selected (that might actually be my fault lol)
- [ ] could probably use on-screen instructions for map controls (click, ctrl+click, etc)
- [ ] toggling compare off and on again half loads what was set - feels like it should remember all or none of it

- [ ] disable comparing to self

- [ ] if content changes while tooltip is up over map, tooltip doesn't update til mouse is moved

    yeah this has annoyed me for a while, but I haven't got around to fixing it. The issue is that we're not setting the tooltips in the view, really.

    I even mention it in the comment at the top of the file :)

    ```
     * Notably, the popups that appear on hover and on click are mostly managed by
     * mapboxgl, which causes some minor problems.
    ```

    Anyway, a fix would be to store the info needed to generate the popups in state and then regenerate them in the mapbox updater.

    Would you be happy to investigate that and fix?

- [ ] selected region in flow view doesn't draw unless flows are toggled off...?

    Don't know what you're talking about here.  See if you can fix it?

- [ ] might be an idea to center chevrons relative to flow length, would lessen stacking when flows overlap

    Could you try this and see? You need to look up the rules for "symbol-spacing" in the mapboxgljs style docs

- [ ] "click to show flows" still appears during a comparison, but clicking disables the comparison

    Can you come up with a better UX for this, maybe just change the tooltip?

- [ ] clicking anywhere on selected regions in flow list deselects, not just the x on the left - is that ok?

    I think this is fine? If you think it's bad UX pls fix

- [ ] region border lines on graph could do with some anti-aliasing

    pls open an issue or fix. v low priority, tho

**Comment by bovine3dom** [comment]
I think the backend is spitting out nonsense data when zones are selected - I'm having a look at it now :)

**Comment by PolyCement** [comment]
>     * [ ]  selected region in flow view doesn't draw unless flows are toggled off...?
>       Don't know what you're talking about here.  See if you can fix it?

Using Boryspil as the example zone selection, what I'm seeing is this with flow lines enabled:
![210707-173828](https://user-images.githubusercontent.com/8337473/124797662-6b8e6480-df4a-11eb-8a11-ece78dfe0a2f.png)

But it's this with flow lines disabled:
![210707-173842](https://user-images.githubusercontent.com/8337473/124797674-6f21eb80-df4a-11eb-9608-d0104d9012d1.png)

Seems like the selected zone should probably be drawn in both cases?

**Comment by cmcaine** [comment]
Yeah, sure. Let's have the outline show at the same time as flow lines.

**Comment by bovine3dom** [comment]
> I think the backend is spitting out nonsense data when zones are selected - I'm having a look at it now :)

I think it's fixed now - Corridors (this repo) shows incoming flows all the time and that's what the UI says too. I've stopped flow lines being shown when everything is zero.

**Comment by PolyCement** [comment]
Currently working on making zone borders show when flow lines are enabled, which means I can demonstrate what I meant by
>     * [ ]  centroids in flow view seem off but again could be bad data

![210708-143735](https://user-images.githubusercontent.com/8337473/124931540-38a3a980-dffa-11eb-8713-fc912f02a0b1.png)
I'm fairly sure the point where all the lines meet is not the centroid of that zone? But it seems more or less correct for a majority of zones so I'm assuming it's just bad data.

**Comment by cmcaine** [comment]
Thanks!

I think we still use Turf.jl to calculate the centroids, so this might be something for Olie or I to look at later. Not a priority.
